drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1434 @@
+
+	/* Connect WU (USB D+) interrupt but make it disabled initially */
+	irq_connect_dynamic(config->wu_irq, 0, it82xx2_wu_isr, dev, 0);
+}
+
+/* The ep_fifo_res[ep_idx % SHARED_FIFO_NUM] where the SHARED_FIFO_NUM is 3 represents the
+ * EP mapping because when (ep_idx % SHARED_FIFO_NUM) is 3, it actually means the EP0.
+ */
+static const uint8_t ep_fifo_res[SHARED_FIFO_NUM] = {3, 1, 2};
+
+static int it82xx2_usb_fifo_ctrl(const struct device *dev, uint8_t ep)
+{
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	volatile uint8_t *ep_fifo_ctrl = usb_regs->fifo_regs[EP_EXT_REGS_BX].fifo_ctrl.ep_fifo_ctrl;
+	uint8_t fifon_ctrl = (ep_fifo_res[ep_idx % SHARED_FIFO_NUM] - 1) * 2;
+	int ret = 0;
+
+	if (ep_idx == 0) {
+		LOG_ERR("Invalid endpoint 0x%x", ep);
+		return -EINVAL;
+	}
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		if (ep_idx < 8) {
+			ep_fifo_ctrl[fifon_ctrl] = BIT(ep_idx);
+			ep_fifo_ctrl[fifon_ctrl + 1] = 0x0;
+		} else {
+			ep_fifo_ctrl[fifon_ctrl] = 0x0;
+			ep_fifo_ctrl[fifon_ctrl + 1] = BIT(ep_idx - 8);
+		}
+	} else if (USB_EP_DIR_IS_OUT(ep)) {
+		if (ep_idx < 8) {
+			ep_fifo_ctrl[fifon_ctrl] |= BIT(ep_idx);
+		} else {
+			ep_fifo_ctrl[fifon_ctrl + 1] |= BIT(ep_idx - 8);
+		}
+	} else {
+		LOG_ERR("Failed to set fifo control register for ep 0x%x", ep);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static void it82xx2_event_submit(const struct device *dev, const uint8_t ep,
+				 const enum it82xx2_event_type event)
+{
+	struct it82xx2_ep_event evt;
+
+	evt.dev = dev;
+	evt.ep = ep;
+	evt.event = event;
+	k_msgq_put(&evt_msgq, &evt, K_NO_WAIT);
+}
+
+static int it82xx2_ep_enqueue(const struct device *dev, struct udc_ep_config *const cfg,
+			      struct net_buf *const buf)
+{
+	udc_buf_put(cfg, buf);
+
+	it82xx2_event_submit(dev, cfg->addr, IT82xx2_EVT_XFER);
+	return 0;
+}
+
+static int it82xx2_ep_dequeue(const struct device *dev, struct udc_ep_config *const cfg)
+{
+	struct net_buf *buf;
+
+	buf = udc_buf_get_all(dev, cfg->addr);
+	if (buf) {
+		udc_submit_ep_event(dev, buf, -ECONNABORTED);
+	}
+
+	udc_ep_set_busy(dev, cfg->addr, false);
+
+	return 0;
+}
+
+static int it82xx2_ep_set_halt(const struct device *dev, struct udc_ep_config *const cfg)
+{
+	const uint8_t ep_idx = USB_EP_GET_IDX(cfg->addr);
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	struct gctrl_it8xxx2_regs *const gctrl_regs = GCTRL_IT8XXX2_REGS_BASE;
+	struct it82xx2_priv_data *priv = udc_get_private(dev);
+
+	if (ep_idx == 0) {
+		unsigned int lock_key;
+		uint32_t idx = 0;
+
+		priv->stall_is_sent = true;
+		lock_key = irq_lock();
+		it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_STALL_SEND, true);
+		it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_READY_ENABLE, true);
+		/* polling if stall send for 3ms */
```

Comment:
Why do you need to poll it?

Reply:
The stall bit may be cleared too late if there are interrupts that happen during setting endpoint0's the stall bit. And it causes the chip responses STALL to the next SETUP transaction. Therefore, after setting stall bit for EP0, the chip polls to check if the stall bit has been transmitted for 3ms and then disables it. 
Besides, the mechanism is following the previous design [usb_dc_ep_set_stall()](https://github.com/zephyrproject-rtos/zephyr/blob/0d27dd5dd3a245df5482d34ccce6a78df5a874b7/drivers/usb/device/usb_dc_it82xx2.c#L1117) in ```usb_dc_it82xx2.c```.

Reply:
The only allowed responses, according to USB specification, to SETUP transaction is ACK or timeout. SETUP transaction cannot be STALLed. If you mean that the next data stage can be STALLed inadvertedly then the solution here is shenanigans. There is no requirement on host to read the STALL within 3 ms or any time really.

Reply:
This still looks wrong and needs further discussion/investigation, I do not have time for this in the next few days, which is a bit of a shame, please ping me if I forget to come back.

Reply:
hi @tmon-nordic @jfischer-no 

It82xx2 does not support clearing the STALL bit by hardware; instead, the STALL bit need to be cleared by firmware. The SETUP token will be STALLed, which does not comply with USB specifications, if firmware clears the STALL bit too late. (e.g., when a transfer is completed, check if it's STALLed and clear STALL bit, shown in the figure) Due to this hardware limitations, the workaround involving polling mechanism is need after setting the STALL.
![image_2024_02_02T03_54_24_295Z](https://github.com/zephyrproject-rtos/zephyr/assets/11677121/070de3ed-d683-47f5-8fed-407df2aa67f0)

Reply:
Thank your for the capture. This seems like a hardware quirk that you have to handle in software. There are device controllers that handle the situation automatically in hardware (which is probably the only way to be fully compliant to the specification).

Reply:
I thinks this block should be moved to `static inline void ctrl_ep_stall_workaround(...){...}` placed above with description you provided.

About workaround self, I am still not entirely convinced. So you lock the cpu and wait for time N, assuming there will be a data/status stage transaction. Why do you assume host will issue a data/status stage within time N? There could be a setup packet instead, very unlikely but possible.

Reply:
Also, in you capture, the host is do not care about STALL handshake right? You can also do `ep_set_halt(dev, USB_CONTROL_EP_OUT, false);...` inside setup interrupt handling.

Reply:
The assumption for data/status stage within time N is probably best what can be done here. By the time the SETUP interrupt comes it is already too late as the device will have already (illegally) responded to SETUP DATA0 with STALL.

Reply:
@Chenhongren Does the driver pass Chapter9 USBCV tests?

Reply:
I think it depends on the host side. In my experiment(the capture I shared), the host's firmware is coreboot. The next transfer is HID get_descriptor and the retry count is 3. [(link)](https://github.com/coreboot/coreboot/blob/cc2ab495256b8fffe4820c768fe1a20be1cd18a7/payloads/libpayload/drivers/usb/usb.c#L171) Consequently, the thransfer succeeds on the next retry. However, without a retry mechanism on the host side, the HID function would fail.
Besides, I have not done the USB20CV test as the requirement states that 'USB20CV can only run on the English language version of Microsoft Windows XP and Microsoft Vista.'[(link)](https://compliance.usb.org/index.asp?UpdateFile=USBCV&Format=Standard#33) Unfortunately, I do not have a computer with this required.

Reply:
Thanks for this information!! The Chapter 9 tests is done and pass.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/11677121/e01ddb00-c19b-4c0f-915b-95b30d7139ad)

Reply:
Also, the workaround part is moved to an inline function.

Reply:
@jfischer-no 
friendly ping